river/cluster/denstream.py

```diff
@@ -43,411 +44,424 @@ class DenStream(base.Clusterer):
+                    # find neighbors of neighbors
+                    neighbor_neighbors = self._query_neighbor(seed_queue[0])
                     # add new neighbors to seed set
                     for neighbor_neighbor in neighbor_neighbors:
                         if labels[neighbor_neighbor] is not None:
-                            seed_set.append(neighbor_neighbor)
+                            seed_queue.append(neighbor_neighbor)
 
-        self.n_clusters, self.clusters = self._generate_clusters_from_labels(labels)
+        self.n_clusters, self.clusters = self._generate_clusters_for_labels(labels)
 
-        y = self._find_closest_cluster_index(x, self.clusters)
-
-        self.centers = {i: self.clusters[i].center for i in self.clusters.keys()}
-
-        return y
+        return self._get_closest_cluster_key(x, self.clusters)
 
 
 class DenStreamMicroCluster(metaclass=ABCMeta):
     """ DenStream Micro-cluster class """
 
-    def __init__(self, x=None, timestamp=None, decaying_factor=None, current_time=None):
+    def __init__(self, x, timestamp, decaying_factor):
 
         self.x = x
-        self.timestamp = timestamp
+        self.last_edit_time = timestamp
         self.creation_time = timestamp
         self.decaying_factor = decaying_factor
-        self.current_time = current_time
-
-        if x is not None and timestamp is not None:
-
-            # initial weight, initial weighted linear sum (IWLS) and initial weighted squared sum (IWSS)
-            # are calculated as described
-
-            self.N = 1
-            self.dim = len(self.x)
-            self.initial_weight = 2 ** (self.decaying_factor * self.timestamp)
-            self.LS = x
-            self.SS = {i: (x[i] * x[i]) for i in range(self.dim)}
-            self.IWLS = {
-                i: (2 ** (self.decaying_factor * self.timestamp)) * x[i]
-                for i in range(self.dim)
-            }
-            self.IWSS = {
-                i: (2 ** (self.decaying_factor * self.timestamp)) * (x[i] * x[i])
-                for i in range(self.dim)
-            }
 
-    @property
-    def weighted_linear_sum(self):
-        weighted_linear_sum = {
-            i: (2 ** (-self.current_time * self.decaying_factor)) * self.IWLS[i]
-            for i in range(self.dim)
-        }
-        return weighted_linear_sum
-
-    @property
-    def weighted_squared_sum(self):
-        weighted_squared_sum = {
-            i: (2 ** (-self.current_time * self.decaying_factor)) * self.IWSS[i]
-            for i in range(self.dim)
-        }
-        return weighted_squared_sum
-
-    @property
-    def weight(self):
-        weight = (
-            2 ** (-self.current_time * self.decaying_factor)
-        ) * self.initial_weight
-        return weight
-
-    @property
-    def center(self):
-        center = {i: self.weighted_linear_sum[i] / self.weight for i in range(self.dim)}
+        self.N = 1
+        self.feature_keys = set(x.keys())
+        self.liner_sum = x
```

Comment:
linear_sum?

Reply:
I just renamed it to follwo the convention in CluStream but I agree that it sounds strange. Basically, we are tracking the sum of values and the sum of squared values. Any suggestion is welcomed.

Reply:
Nice catch. Fixed in c7ec6f9a8ae5cb5ffdbf5d9d6630c6105d8cfc60